jx promote should validate input and fail gracefully

When trying to execute the `jx promote` command, I stumbled upon the following issue:

```
~/personal/repos/jxtest ⑂wip  jx promote
Promoting latest version of app jxtest to namespace jx
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x120 pc=0x21554ad]

goroutine 1 [running]:
github.com/jenkins-x/jx/pkg/jx/cmd.(*PromoteOptions).createPromoteKey(0xc420001500, 0x0, 0x8)
	/home/jenkins/go/src/github.com/jenkins-x/jx/pkg/jx/cmd/promote.go:872 +0x127d
github.com/jenkins-x/jx/pkg/jx/cmd.(*PromoteOptions).Promote(0xc420001500, 0xc42031b7b8, 0x2, 0x0, 0x1, 0xc420068cac, 0x6, 0xc42031a920)
	/home/jenkins/go/src/github.com/jenkins-x/jx/pkg/jx/cmd/promote.go:310 +0x425
github.com/jenkins-x/jx/pkg/jx/cmd.(*PromoteOptions).Run(0xc420001500, 0xc4202e3c50, 0x0)
	/home/jenkins/go/src/github.com/jenkins-x/jx/pkg/jx/cmd/promote.go:213 +0x286
github.com/jenkins-x/jx/pkg/jx/cmd.NewCmdPromote.func1(0xc42021b400, 0x31fab00, 0x0, 0x0)
	/home/jenkins/go/src/github.com/jenkins-x/jx/pkg/jx/cmd/promote.go:118 +0x75
github.com/jenkins-x/jx/vendor/github.com/spf13/cobra.(*Command).execute(0xc42021b400, 0x31fab00, 0x0, 0x0, 0xc42021b400, 0x31fab00)
	/home/jenkins/go/src/github.com/jenkins-x/jx/vendor/github.com/spf13/cobra/command.go:760 +0x2c1
github.com/jenkins-x/jx/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc420094c80, 0xc4202d7180, 0xc4203c1680, 0xc4203bf400)
	/home/jenkins/go/src/github.com/jenkins-x/jx/vendor/github.com/spf13/cobra/command.go:846 +0x334
github.com/jenkins-x/jx/vendor/github.com/spf13/cobra.(*Command).Execute(0xc420094c80, 0xc420298f18, 0x309e300)
	/home/jenkins/go/src/github.com/jenkins-x/jx/vendor/github.com/spf13/cobra/command.go:794 +0x2b
github.com/jenkins-x/jx/cmd/jx/app.Run(0xc4204a9f70, 0x2185ec9)
	/home/jenkins/go/src/github.com/jenkins-x/jx/cmd/jx/app/jx.go:18 +0x90
main.main()
	/home/jenkins/go/src/github.com/jenkins-x/jx/cmd/jx/jx.go:10 +0x22
```

## Environment
```
NAME           VERSION
jx             1.1.42
Jenkins X      0.0.543
Kubernetes     v1.8.8-gke.0
Helm Client    v2.8.2+ga802316
Helm Server    v2.8.2+ga802316
Kubectl Client v1.9.0
Git            git version 2.15.1
```

When executing the command with the correct arguments: `jx promote --version v0.0.2 --env production --timeout 1h`, the execution is successful.

`jx promote` could validate the user input and print a descriptive help message if the input is not valid.

Associated with https://github.com/jenkins-x/jx/issues/357, with the difference that this is not a bug in the command, rather than not enough checks being made to prevent from erroneous user input / graceful failure.

--- a/pkg/addon/config.go
+++ b/pkg/addon/config.go
@@ -9,7 +9,6 @@ import (
 	"gopkg.in/yaml.v2"
 )
 
-
 //AddonConfig Addon Configration
 type AddonConfig struct {
 	Name    string

--- a/pkg/gits/bitbucket.go
+++ b/pkg/gits/bitbucket.go
@@ -503,7 +503,11 @@ func (b *BitbucketProvider) JenkinsWebHookPath(gitURL string, secret string) str
 }
 
 func (b *BitbucketProvider) Label() string {
-	return ""
+	return b.Server.Label()
+}
+
+func (b *BitbucketProvider) ServerURL() string {
+	return b.Server.URL
 }
 
 func (b *BitbucketProvider) UpdateRelease(owner string, repo string, tag string, releaseInfo *GitRelease) error {

--- a/pkg/gits/gitea.go
+++ b/pkg/gits/gitea.go
@@ -531,3 +531,7 @@ func GiteaAccessTokenURL(url string) string {
 func (p *GiteaProvider) Label() string {
 	return p.Server.Label()
 }
+
+func (p *GiteaProvider) ServerURL() string {
+	return p.Server.URL
+}

--- a/pkg/gits/github.go
+++ b/pkg/gits/github.go
@@ -662,6 +662,10 @@ func (p *GitHubProvider) Label() string {
 	return p.Server.Label()
 }
 
+func (p *GitHubProvider) ServerURL() string {
+	return p.Server.URL
+}
+
 func asBool(b *bool) bool {
 	if b != nil {
 		return *b

--- a/pkg/gits/gitlab.go
+++ b/pkg/gits/gitlab.go
@@ -395,6 +395,10 @@ func (g *GitlabProvider) Label() string {
 	return g.Server.Label()
 }
 
+func (p *GitlabProvider) ServerURL() string {
+	return p.Server.URL
+}
+
 func (g *GitlabProvider) UpdateRelease(owner string, repo string, tag string, releaseInfo *GitRelease) error {
 	return nil
 }

--- a/pkg/gits/provider.go
+++ b/pkg/gits/provider.go
@@ -88,7 +88,11 @@ type GitProvider interface {
 
 	JenkinsWebHookPath(gitURL string, secret string) string
 
+	// Label returns the git service label or name
 	Label() string
+
+	// ServerURL returns the git server URL
+	ServerURL() string
 }
 
 type GitOrganisation struct {

--- a/pkg/issues/constants.go
+++ b/pkg/issues/constants.go
@@ -4,6 +4,7 @@ const (
 	Bugzilla = "bugzilla"
 	Jira     = "jira"
 	Trello   = "trello"
+	Git      = "git"
 )
 
 var (

--- a/pkg/issues/git_provider.go
+++ b/pkg/issues/git_provider.go
@@ -5,6 +5,7 @@ import (
 	"strconv"
 
 	"github.com/jenkins-x/jx/pkg/gits"
+	"github.com/jenkins-x/jx/pkg/util"
 )
 
 type GitIssueProvider struct {
@@ -65,4 +66,8 @@ func (i *GitIssueProvider) CreateIssueComment(key string, comment string) error 
 		return err
 	}
 	return i.GitProvider.CreateIssueComment(i.Owner, i.Repository, n, comment)
+}
+
+func (i *GitIssueProvider) HomeURL() string {
+	return util.UrlJoin(i.GitProvider.ServerURL(), i.Owner, i.Repository)
 }

--- a/pkg/issues/jira.go
+++ b/pkg/issues/jira.go
@@ -107,6 +107,40 @@ func (i *JiraService) jiraToGitIssue(issue *jira.Issue) *gits.GitIssue {
 		answer.Body = fields.Description
 		answer.Labels = gits.ToGitLabels(fields.Labels)
 		answer.ClosedAt = jiraTimeToTimeP(fields.Resolutiondate)
+		answer.User = jiraUserToGitUser(fields.Reporter)
+		assignee := jiraUserToGitUser(fields.Assignee)
+		if assignee != nil {
+			answer.Assignees = []gits.GitUser{*assignee}
+		}
+	}
+	return answer
+}
+
+func jiraUserToGitUser(user *jira.User) *gits.GitUser {
+	if user == nil {
+		return nil
+	}
+	return &gits.GitUser{
+		AvatarURL: jiraAvatarUrl(user),
+		Name:      user.Name,
+		Login:     user.Key,
+		Email:     user.EmailAddress,
+	}
+}
+func jiraAvatarUrl(user *jira.User) string {
+	answer := ""
+	if user != nil {
+		av := user.AvatarUrls
+		answer = av.Four8X48
+		if answer == "" {
+			answer = av.Three2X32
+		}
+		if answer == "" {
+			answer = av.Two4X24
+		}
+		if answer == "" {
+			answer = av.One6X16
+		}
 	}
 	return answer
 }
@@ -134,4 +168,8 @@ func (i *JiraService) gitToJiraIssue(issue *gits.GitIssue) *jira.Issue {
 
 func (i *JiraService) ServerName() string {
 	return i.Server.URL
+}
+
+func (i *JiraService) HomeURL() string {
+	return util.UrlJoin(i.Server.URL, "browse", i.Project)
 }

--- a/pkg/issues/provider.go
+++ b/pkg/issues/provider.go
@@ -22,6 +22,9 @@ type IssueProvider interface {
 
 	// IssueURL returns the URL of the given issue for this project
 	IssueURL(key string) string
+
+	// HomeURL returns the home URL of the issue tracker
+	HomeURL() string
 }
 
 func CreateIssueProvider(kind string, server *auth.AuthServer, userAuth *auth.UserAuth, project string) (IssueProvider, error) {
@@ -41,4 +44,13 @@ func ProviderAccessTokenURL(kind string, url string) string {
 	default:
 		return ""
 	}
+}
+
+// GetIssueProvider returns the kind of issue provider
+func GetIssueProvider(tracker IssueProvider) string {
+	_, ok := tracker.(*JiraService)
+	if ok {
+		return Jira
+	}
+	return Git
 }

--- a/pkg/jx/cmd/import.go
+++ b/pkg/jx/cmd/import.go
@@ -200,10 +200,6 @@ func (o *ImportOptions) Run() error {
 			return err
 		}
 	}
-	if o.Organisation == "" {
-		o.Organisation = userAuth.Username
-	}
-
 	if o.GitHub {
 		return o.ImportProjectsFromGitHub()
 	}

--- a/pkg/jx/cmd/promote.go
+++ b/pkg/jx/cmd/promote.go
@@ -210,6 +210,12 @@ func (o *PromoteOptions) Run() error {
 	if o.AllAutomatic {
 		return o.PromoteAllAutomatic()
 	}
+	if env == nil {
+		if o.Environment == "" {
+			return util.MissingOption(optionEnvironment)
+		}
+		return fmt.Errorf("Could not find an Environment called %s", o.Environment)
+	}
 	releaseInfo, err := o.Promote(targetNS, env, true)
 	err = o.WaitForPromotion(targetNS, env, releaseInfo)
 	if err != nil {
@@ -868,7 +874,7 @@ func (o *PromoteOptions) createPromoteKey(env *v1.Environment) *kube.PromoteStep
 	releaseName := o.ReleaseName
 	if o.releaseResource == nil && releaseName != "" {
 		jxClient, _, err := o.JXClient()
-		if err == nil {
+		if err == nil && jxClient != nil {
 			release, err := jxClient.JenkinsV1().Releases(env.Spec.Namespace).Get(releaseName, metav1.GetOptions{})
 			if err == nil && release != nil {
 				o.releaseResource = release

--- a/pkg/jx/cmd/status.go
+++ b/pkg/jx/cmd/status.go
@@ -96,16 +96,16 @@ func (o *StatusOptions) Run() error {
 	resourceStr := clusterStatus.CheckResource()
 	jenkinsURL, err := o.findServiceInNamespace("jenkins", namespace)
 	if err != nil {
-		if resourceStr != ""{
-			log.Warnf("%s Jenkins not found and %s\n", clusterStatus.Info(),resourceStr)
-		}else{
+		if resourceStr != "" {
+			log.Warnf("%s Jenkins not found and %s\n", clusterStatus.Info(), resourceStr)
+		} else {
 			log.Warnf("%s Jenkins not found\n", clusterStatus.Info())
 		}
 		return err
 	}
 	if resourceStr != "" {
-		log.Warnf("Jenkins X installed for %s. Jenkins is running at %s. Need more %s\n", clusterStatus.Info(), jenkinsURL,resourceStr)
-	}else{
+		log.Warnf("Jenkins X installed for %s. Jenkins is running at %s. Need more %s\n", clusterStatus.Info(), jenkinsURL, resourceStr)
+	} else {
 		log.Successf("Jenkins X checks passed for %s. Jenkins is running at %s\n", clusterStatus.Info(), jenkinsURL)
 	}
 

--- a/pkg/jx/cmd/step_changelog.go
+++ b/pkg/jx/cmd/step_changelog.go
@@ -9,7 +9,6 @@ import (
 	"os"
 	"path/filepath"
 	"regexp"
-	"strconv"
 	"strings"
 	"text/template"
 	"time"
@@ -18,6 +17,7 @@ import (
 	"github.com/jenkins-x/jx/pkg/apis/jenkins.io"
 	"github.com/jenkins-x/jx/pkg/apis/jenkins.io/v1"
 	"github.com/jenkins-x/jx/pkg/gits"
+	"github.com/jenkins-x/jx/pkg/issues"
 	"github.com/jenkins-x/jx/pkg/jx/cmd/templates"
 	"github.com/jenkins-x/jx/pkg/kube"
 	"github.com/jenkins-x/jx/pkg/util"
@@ -54,6 +54,7 @@ type StepChangelogOptions struct {
 type StepChangelogState struct {
 	GitInfo         *gits.GitRepositoryInfo
 	GitProvider     gits.GitProvider
+	Tracker         issues.IssueProvider
 	FoundIssueNames map[string]bool
 }
 
@@ -124,6 +125,7 @@ e.g. define environment variables GIT_USERNAME and GIT_API_TOKEN
 `)
 
 	GitHubIssueRegex = regexp.MustCompile(`(\#\d+)`)
+	JIRAIssueRegex   = regexp.MustCompile(`[A-Z][A-Z]+-(\d+)`)
 )
 
 func NewCmdStepChangelog(f cmdutil.Factory, out io.Writer, errOut io.Writer) *cobra.Command {
@@ -243,6 +245,12 @@ func (o *StepChangelogOptions) Run() error {
 		return err
 	}
 	o.State.GitInfo = gitInfo
+
+	tracker, err := o.createIssueProvider(o.Dir)
+	if err != nil {
+		return err
+	}
+	o.State.Tracker = tracker
 
 	authConfigSvc, err := o.Factory.CreateGitAuthConfigService()
 	if err != nil {
@@ -426,58 +434,54 @@ func (o *StepChangelogOptions) addCommit(spec *v1.ReleaseSpec, commit *object.Co
 }
 
 func (o *StepChangelogOptions) addIssuesAndPullRequests(spec *v1.ReleaseSpec, commit *v1.CommitSummary) error {
-	// TODO for now assume github but allow a JIRA server & project name to be specified...
+	tracker := o.State.Tracker
 
 	gitProvider := o.State.GitProvider
 	if gitProvider == nil || !gitProvider.HasIssues() {
 		return nil
 	}
-	gitInfo := o.State.GitInfo
-	matches := GitHubIssueRegex.FindAllStringSubmatch(commit.Message, -1)
+	regex := GitHubIssueRegex
+	if issues.GetIssueProvider(tracker) == issues.Jira {
+		regex = JIRAIssueRegex
+	}
+	matches := regex.FindAllStringSubmatch(commit.Message, -1)
 	for _, match := range matches {
 		for _, result := range match {
-			numberText := strings.TrimPrefix(result, "#")
+			result = strings.TrimPrefix(result, "#")
 			if _, ok := o.State.FoundIssueNames[result]; !ok {
 				o.State.FoundIssueNames[result] = true
-				number, err := strconv.Atoi(numberText)
+				issue, err := tracker.GetIssue(result)
 				if err != nil {
-					o.warnf("Failed to convert issue reference %s into a number %s\n", numberText, err)
-					continue
-				}
-				owner := gitInfo.Organisation
-				repo := gitInfo.Name
-				issue, err := gitProvider.GetIssue(owner, repo, number)
-				if err != nil {
-					o.warnf("Failed to lookup issue %s in repository\n", result, err)
+					o.warnf("Failed to lookup issue %s in issue tracker %s due to %s\n", result, tracker.HomeURL(), err)
 					continue
 				}
 				if issue == nil {
-					o.warnf("Failed to find issue %d for repository %s/%s\n", number, owner, repo)
+					o.warnf("Failed to find issue %s for repository %s\n", result, tracker.HomeURL())
 				}
 
 				var user v1.UserDetails
 				if issue.User == nil {
-					o.warnf("Failed to find user for issue %d repository %s/%s\n", number, owner, repo)
+					o.warnf("Failed to find user for issue %s repository %s\n", result, tracker.HomeURL())
 				} else {
 					user = *o.gitUserToUserDetails(issue.User)
 				}
 
 				var closedBy v1.UserDetails
 				if issue.ClosedBy == nil {
-					o.warnf("Failed to find closedBy user for issue %d repository %s/%s\n", number, owner, repo)
+					o.warnf("Failed to find closedBy user for issue %s repository %s\n", result, tracker.HomeURL())
 				} else {
 					closedBy = *o.gitUserToUserDetails(issue.User)
 				}
 
 				var assignees []v1.UserDetails
 				if issue.Assignees == nil {
-					o.warnf("Failed to find assignees for issue %d repository %s/%s\n", number, owner, repo)
+					o.warnf("Failed to find assignees for issue %s repository %s\n", result, tracker.HomeURL())
 				} else {
 					assignees = o.gitUserToUserDetailSlice(issue.Assignees)
 				}
 
 				issueSummary := v1.IssueSummary{
-					ID:        numberText,
+					ID:        result,
 					URL:       issue.URL,
 					Title:     issue.Title,
 					Body:      issue.Body,

--- a/pkg/jx/cmd/step_changelog.go
+++ b/pkg/jx/cmd/step_changelog.go
@@ -246,7 +246,7 @@ func (o *StepChangelogOptions) Run() error {
 	}
 	o.State.GitInfo = gitInfo
 
-	tracker, err := o.createIssueProvider(o.Dir)
+	tracker, err := o.createIssueProvider(dir)
 	if err != nil {
 		return err
 	}

--- a/pkg/kube/cluster_status.go
+++ b/pkg/kube/cluster_status.go
@@ -106,14 +106,12 @@ func (clusterStatus *ClusterStatus) NodeCount() int {
 	return clusterStatus.nodeCount
 }
 
-
-
-func (clusterStatus *ClusterStatus) CheckResource()  string {
+func (clusterStatus *ClusterStatus) CheckResource() string {
 	if clusterStatus.AverageMemPercent() >= clusterStatus.MinimumResourceLimit() {
-		return  "needs more free memory"
+		return "needs more free memory"
 	}
 	if clusterStatus.AverageCpuPercent() >= clusterStatus.MinimumResourceLimit() {
-		return  "needs more free compute"
+		return "needs more free compute"
 	}
 	return ""
 }
@@ -128,7 +126,6 @@ func (clusterStatus *ClusterStatus) Info() string {
 		clusterStatus.totalAllocatedCpu.String())
 	return str
 }
-
 
 func Status(client *kubernetes.Clientset, namespace string, node v1.Node) (NodeStatus, error) {
 	nodeStatus := NodeStatus{}

